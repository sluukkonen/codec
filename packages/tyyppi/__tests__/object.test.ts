import { fromJson, number, object, optional, string } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

test("should parse an empty object", () => {
  expectParseSuccess(object({}), {})
})

test("should fail to parse non-objects", () => {
  expectParseFailure(object({}), "")
})

test("should parse a valid object", () => {
  expectParseSuccess(object({ a: number, b: string }), { a: 1, b: "1" })
})

test("should parse complex object", () => {
  expectParseSuccess(object({ n: fromJson.bigint }), { n: "1" }, { n: 1n })
})

test("should fail to parse an object with the value", () => {
  expectParseFailure(object({ a: string }), { a: 1 })
})

test("should work with nested objects", () => {
  expectParseSuccess(object({ a: object({ b: string }) }), { a: { b: "" } })
})

test("the path property should work with nested objects", () => {
  expectParseFailure(object({ a: object({ b: string }) }), { a: { b: 0 } })
})

test("should not write undefined properties", () => {
  const codec = object({ a: fromJson.bigint, b: optional(string) })
  expect(codec.decodeOrThrow({ a: "1" })).toEqual({ a: 1n })
  expect(codec.encode({ a: 1n })).toStrictEqual({ a: "1" })
})

test("should not accept null values", () => {
  expectParseFailure(object({ a: string }), null)
})

test("should not accept arrays", () => {
  expectParseFailure(object({ "0": number }), [0])
})

test("should only consider own properties", () => {
  const obj = Object.assign(Object.create({ a: 1 }), { b: 2, c: 3 })
  expectParseFailure(object({ a: number, b: number, c: number }), obj)
})

test("should not strip extra keys (simple)", () => {
  expectParseSuccess(object({}), { a: 1, b: 2 })
})

test("should not strip extra keys (complex)", () => {
  expectParseSuccess(
    object({ n: fromJson.bigint }),
    { n: "1", a: 1, b: 2 } as { n: string },
    { n: 1n, a: 1, b: 2 } as { n: bigint },
  )
})
