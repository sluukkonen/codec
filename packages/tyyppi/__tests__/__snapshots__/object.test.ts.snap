// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail to parse an object with the value 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected a string, received a number",
    "path": [
      "a",
    ],
    "received": "number",
  },
]
`;

exports[`should fail to parse non-objects 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Expected an object, received a string",
    "path": [],
    "received": "string",
  },
]
`;

exports[`should not accept arrays 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Expected an object, received an array",
    "path": [],
    "received": "array",
  },
]
`;

exports[`should not accept null values 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Expected an object, received null",
    "path": [],
    "received": "null",
  },
]
`;

exports[`should only consider own properties 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected a number, received undefined",
    "path": [
      "a",
    ],
    "received": "undefined",
  },
]
`;

exports[`the path property should work with nested objects 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected a string, received a number",
    "path": [
      "a",
      "b",
    ],
    "received": "number",
  },
]
`;
